kumascript/src/info.js

```diff
@@ -1,149 +1,55 @@
-      if (alreadyProcessed.has(childUri)) {
-        return;
-      }
-      alreadyProcessed.add(childUri);
-      slugSegments.pop();
-      const parentUri = `/${locale}/docs/${slugSegments.join("/")}`;
-      if (pagesByUri.has(childUri)) {
-        // This URI is an actual document, so add it as a child of
-        // its parent URI.
-        if (childrenByUri.has(parentUri)) {
-          childrenByUri.get(parentUri).push(childUri);
-        } else {
-          childrenByUri.set(parentUri, [childUri]);
-        }
-      }
-      if (slugSegments.length > 1) {
-        addChild(parentUri, locale, slugSegments);
-      }
-    }
-
-    // Build "childrenByUri", a map of every possible parent URI
-    // to its list of actual child document URI's.
-    for (const uri of pagesByUri.keys()) {
-      const [locale, ...slugSegments] = uri
-        .replace("/docs/", " ")
-        .replace(/\//g, " ")
-        .trim()
-        .split(" ");
-      addChild(uri, locale, slugSegments);
+function repairURL(url) {
+  // Returns a lowercase URI with common irregularities repaired.
+  url = url.trim().toLowerCase();
+  if (!url.startsWith("/")) {
+    // Ensure the URI starts with a "/".
+    url = "/" + url;
+  }
+  // Remove redundant forward slashes, like "//".
+  url = url.replace(/\/{2,}/g, "/");
+  // Ensure the URI starts with a valid locale.
+  const maybeLocale = url.split("/")[1];
+  if (!VALID_LOCALES.has(maybeLocale)) {
+    if (maybeLocale === "en") {
+      // Converts URI's like "/en/..." to "/en-us/...".
+      url = url.replace(`/${maybeLocale}`, "/en-us");
+    } else {
+      // Converts URI's like "/web/..." to "/en-us/web/...", or
+      // URI's like "/docs/..." to "/en-us/docs/...".
+      url = "/en-us" + url;
     }
   }
+  // Ensure the locale is followed by "/docs".
+  const [locale, maybeDocs] = url.split("/").slice(1, 3);
+  if (maybeDocs !== "docs") {
+    // Converts URI's like "/en-us/web/..." to "/en-us/docs/web/...".
+    url = url.replace(`/${locale}`, `/${locale}/docs`);
+  }
+  return url;
+}
 
+const info = {
   getPathname(url) {
     // This function returns just the pathname of the given "url", removing
     // any trailing "/".
     return new URL(url, DUMMY_BASE_URL).pathname.replace(/\/$/, "");
-  }
+  },
 
   getUriKey(url) {
     // This function returns just the lowercase pathname of the given "url",
-    // but potentially transformed (i.e., repaired and/or redirected).
-    const uri = this.getPathname(url);
-    return this.uriTransform(uri.toLowerCase());
-  }
+    // removing any trailing "/". The BASE_URL is not important here, since
+    // we're only after the path of any incoming "url", but it's required by
+    // the URL constructor when the incoming "url" is relative.
+    const uri = new URL(url, BASE_URL).pathname
+      .replace(/\/$/, "")
+      .toLowerCase();
+    return Redirect.resolve(repairURL(uri));
+  },
 
   getDescription(url) {
     const uriKey = this.getUriKey(url);
```

Comment:
Is it safer to use `info.getUriKey` instead?

Reply:
Yes, much easier on the eyes. Thanks for the suggestion